docs/architecture-decision-records/2023-03-integration-with-station.md

```diff
@@ -0,0 +1,232 @@
+
+   As a CLI tool, it reads the configuration from a config file (either user-level or
+   project-specific) or CLI arguments. There should be reasonable defaults tailored to developers
+   building a module - e.g. the state files should be stored in the project's working directory.
+
+2. A deployment platform, running modules inside Filecoin Station instances operated by
+   non-technical users. We want to optimize for a smooth user experience for Station operators
+   first, and easy troubleshooting of operations by module developers second.
+
+   In this mode, Zinnia is a long-running process executing multiple modules. It needs to support
+   reloading individual modules when a new version of a module is deployed. It also needs to report
+   various information back to the Station, e.g. activity log and the number of jobs completed.
+
+   As a service, it should read the configuration from environment variables (see
+   [The Twelve-Factor App](https://12factor.net/config)) and allow embedders to improve security
+   e.g. by keeping modules in a different place from the runtime state.
+
+### `zinniad`
+
+For the reasons above, I propose to build a new binary called `zinniad` that will execute Zinnia
+modules inside the Station.
+
+_(This is a strawman proposal, subject to changes during implementation.)_
+
+```bash
+# Address of Station's built-in wallet
+FIL_WALLET_ADDRESS=f1etc \
+# Where to keep state files
+STATE_ROOT=$HOME/Library/Caches/Filecoin\ Station/zinnia \
+# The program to run. The only argument is path to the directory containing module files.
+zinniad /Applications/Filecoin\ Station.app/Contents/Resources/zinnia-modules
+```
+
+### Communication with Station (Core, Desktop):
+
+As explained above, I propose to use environment variables to pass configuration from the Station to
+Zinnia.
+
+For communicating information from Zinnia back to the Station, I propose that Zinnia prints
+newline-delimited JSON entries to `stdout` and uses `stderr` for general/debug logging. This format
+is easy to parse from Node.js code powering both Station variants.
+
+Example messages:
+
+- **Activity log - error**
+
+  `{"type": "activity:error", "module": "saturn", "message": "Cannot connect to the orchestrator."x }`
+
+  \_Note: `"module": "saturn"` describes which module emitted the log.
+
+- **Activity log - info**
+
+  `{"type": "activity:info", "module": null, "message": "Zinnia is starting up..."}`
+
+  _Note: `"module": null` means the message comes from Zinnia runtime._
+
+- **Number of jobs completed:**
+
+  `{"type": "jobs-completed", "total": 123 }`
+
+  _Note: This message is emitted periodically, e.g. every 200ms._
+
+  In the future, we can easily extend this line to include per-module stats too:
+
+  `{"type": "jobs-completed", "total": 123, modules: {"saturn": 100, "retrieval-checker": 23}}`
+
+Discussion points:
+
+- Should these logs include a timestamp? As a new JSON field or as a prefix before the JSON starts?
```

Comment:
If the consumer sits on the same machine, it shouldn't matter who will attach the timestamp. If consumers sit on different machines (with clock drift), then it would be better if Zinnia attached a timestamp, acting as the source of trough.

Reply:
This a great way how to look at this problem 💪🏻 

Our current goal is integration with the Station. The Station (Core or Desktop) and the Zinnia runtime will initially sit on the same machine.

My conclusion is not to add timestamps to the logs now. We can easily add them later when a need arises.

Reply:
See [71994ec](https://github.com/filecoin-station/zinnia/pull/121/commits/71994ec84e91e380c45005db6330df596aecd45c)